net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+
+		n = cp - name;
+		if (n == 0 )
+			return M0_ERR(-EINVAL);
+
+		++cp;
+	}
+
+	if ((nodeSize < (n+1)) || (portSize < (strlen(cp)+1)))
+		return M0_ERR(-EINVAL);
+
+	memcpy(node, name, n);
+	node[n] = '\0';
+
+	n=strlen(cp);
+	memcpy(port, cp, n);
+	port[n] = '\0';
+
+	return M0_RC(rc);
+}
+
+/**
+ * Parses network address.
+ *
+ * The following address formats are supported:
+ *
+ *     - lnet compatible, see nlx_core_ep_addr_decode():
+ *
+ *           nid:pid:portal:tmid
+ *
+ *       for example: "10.0.2.15@tcp:12345:34:123" or
+ *       "192.168.96.128@tcp1:12345:31:*"
+ *
+ *     - sock format, see socket(2):
+ *           family:type:ipaddr[@port]
+ *
+ *     - libfab compatible format
+ *       for example IPV4 libfab:192.168.0.1:4235
+ *                   IPV6 libfab:[4002:db1::1]:4235
+ *
+ */
+static int libfab_ep_addr_decode(struct m0_fab__ep *ep, const char *name)
+{
+	char *node = ep->fep_name.fen_addr;
+	char *port = ep->fep_name.fen_port;
+	size_t nodeSize = ARRAY_SIZE(ep->fep_name.fen_addr);
+	size_t portSize = ARRAY_SIZE(ep->fep_name.fen_port);
+	int result;
+
+	M0_ENTRY("name=%s", name);
+	
+	if( name == NULL || name[0] == 0)
+		result =  M0_ERR(-EPROTO);
+	else if ((strncmp(name, "libfab", 6)) == 0)
+		result = libfab_ep_addr_decode_native(name, node, nodeSize, 
+						      port, portSize);
+	else if (name[0] < '0' || name[0] > '9')
+		/* sock format */
+		result = libfab_ep_addr_decode_sock(name, node, nodeSize, 
+						    port, portSize);
+	else
+		/* Lnet format. */
+		result = libfab_ep_addr_decode_lnet(name, node, nodeSize, 
+						    port, portSize);
+
+	if (result == FI_SUCCESS)
+		strcpy(ep->fep_name.fen_str_addr, name);
+
+	return M0_RC(result);
+}
+
+/**
+ * Used to lock the transfer machine mutex
+ */
+static inline void libfab_tm_lock(struct m0_fab__tm *tm)
+{
+	m0_mutex_lock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to unlock the transfer machine mutex
+ */
+static inline void libfab_tm_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_ntm->ntm_mutex);
+}
+
+static inline int libfab_tm_trylock(struct m0_fab__tm *tm)
+{
+	return m0_mutex_trylock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to lock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_lock(struct m0_fab__tm *tm)
+{
+	m0_mutex_lock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to unlock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to check if the transfer machine mutex is locked.
+ * Returns true if locked, else false.
+ */
+static inline bool libfab_tm_is_locked(const struct m0_fab__tm *tm)
+{
+	return m0_mutex_is_locked(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Helper function that posts a tm state change event.
+ */
+static void libfab_tm_event_post(struct m0_fab__tm *tm,
+				 enum m0_net_tm_state state)
+{
+	struct m0_net_end_point *listen = NULL;
+
+	if (state == M0_NET_TM_STARTED) {
```

Comment:
Do we need to check this (If TM started, listening endpoint should be there) all time for each event post?
If TM is not started and listen is NULL is this valid scenario? Event post possible in this scenario?

Reply:
libfab_tm_event_post() is used only for posting the M0_NET_TM_STARTED and the M0_NET_TM_STOPPED events and in case of M0_NET_TM_STOPPED, the condition (TM is not started and listen is NULL) is valid.